### DISCUSSION TITLE
Speed up soft shadows

### DISCUSSION THREAD

**Discussion by kpeeters** [google/filament#5039] [General]
Since PR #4817 went in, soft shadows can no longer be obtained using PCF. However, VSM shadows seem to be substantially slower. Any suggestions for a good set of VSM parameters that gets me reasonably close to the PCF soft shadow quality, but does not take a lot more computational power? (aiming at mobile).

**Comment by romainguy** [comment]
@pixelflinger

**Comment by pixelflinger** [comment]
It's surprising that VSM would be substantially slower -- it wasn't the case in my tests. VSM achieves much larger blurs much quicker. How much slower is it for you and on which hardware?

For VSM, make sure to not generate mipmaps and to not use anisotropic filtering. If you want anti-aliased shadows (as opposed to soft), the best setting is MSAA 4x and no blur. If you want a fair amount of softness, I would recommend turning of MSAA and setting the blur to your desired value. Larger blurs don't cost much more once blurring is enabled, however too large blurs cause light leaking artifacts.

**Comment by kpeeters** [comment]
The results below are for a Samsung Tab S6 (Adreno 640), but I see similar things on other devices. The scenes I am rendering are generated by rather elaborate code which I don't want to distract you with, but I'll can try to export some static elements of it to GLTF if necessary.

The scene contains about 85k triangles, 2 directional lights (one casting shadows) and a spotlight (also casting shadows). The choreographer is verified to run at 60fps. My per-frame CPU code never takes more than 10ms. With PCF shadows, in about 30% of frames `beginFrame` returns `false` (which is a bit high, so maybe the key is there). If I switch to VSM shadows, `beginFrame` returns `false` on 80% of the frames. 

This is all with dynamic resolution turned on, 4x MSAA (`view->setSampleCount(4)`), and a VSM blur width of 4 when enabled.

**Comment by kpeeters** [comment]
The triangles are distributed over about 100 buffers, maybe that is important. Most of them share the same material instance though.

**Comment by pixelflinger** [comment]
And you do not have any other setting enabled for VSM (especially MSAA).

**Reply by kpeeters** [reply]
Everything else for VSM is on default settings.

**Comment by pixelflinger** [comment]
Okay that is interesting, I wasn't expecting that. The old PCF was doing a 3x3 gaussian filter, which would be equivalent to VSM + blur=3. But it was optimized to do only 4 taps.
So, yeah in fact it's probably true that it would beat VSM + blur.

Okay, I'm going to rethink that for a bit :-)

**Reply by kpeeters** [reply]
Thanks, much appreciated!

**Reply by pixelflinger** [reply]
I put back the 3x3 filter on PCF.
